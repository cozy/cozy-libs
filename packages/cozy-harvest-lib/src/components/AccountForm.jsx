import React, { PureComponent } from 'react'
import { Form, Field as FinalFormField } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'
import Field from 'cozy-ui/react/Field'

import AccountFormError from './AccountFormError'
import {
  getEncryptedFieldName,
  getFieldPlaceholder,
  sanitizeSelectProps
} from '../helpers/fields'
import { KonnectorJobError } from '../helpers/konnectors'
import manifest from 'helpers/manifest'
import OAuthForm from './OAuthForm'

const predefinedLabels = [
  'answer',
  'birthdate',
  'code',
  'date',
  'email',
  'firstname',
  'lastname',
  'login',
  'password',
  'phone'
]

// Out of scope labels already used, should be transferred directly in manifests
// in the future.
const legacyLabels = [
  'branchName' // Used in banking konnectors
]

const VALIDATION_ERROR_REQUIRED_FIELD = 'VALIDATION_ERROR_REQUIRED_FIELD'

export class AccountField extends PureComponent {
  render() {
    const {
      disabled,
      hasError,
      initialValue,
      label,
      name,
      onKeyUp,
      required,
      role,
      t,
      type
    } = this.props

    // Allow manifest to specify predefined label
    const localeKey =
      predefinedLabels.includes(label) || legacyLabels.includes(label)
        ? label
        : name

    const isEditable = !(role === 'identifier' && initialValue)

    const fieldProps = {
      ...this.props,
      autoComplete: 'off',
      className: 'u-m-0', // 0 margin
      disabled: disabled || !isEditable,
      error: hasError,
      fullwidth: true,
      label: t(`fields.${localeKey}.label`, {
        _: t(`legacy.fields.${localeKey}.label`, { _: name })
      }),
      onKeyUp: onKeyUp,
      placeholder: getFieldPlaceholder(
        this.props,
        t(`fields.${name}.placeholder`, { _: '' })
      ),
      side: !required && t('accountForm.fields.optional'),
      size: 'medium'
    }
    const passwordLabels = {
      hideLabel: t('accountForm.password.hide'),
      showLabel: t('accountForm.password.show')
    }
    switch (type) {
      case 'date':
        return (
          <Field
            {...fieldProps}
            placeholder={t(`fields.${name}.placeholder`, {
              _: t(`default.dateFormat`).toLowerCase()
            })}
          />
        )
      case 'dropdown':
        return <Field {...sanitizeSelectProps(fieldProps)} />
      case 'password':
        return (
          <Field
            {...fieldProps}
            // Usting the `new-password` value is the best way to avoid
            // autocomplete for password.
            // See https://stackoverflow.com/a/17721462/1135990
            autoComplete="new-password"
            secondaryLabels={passwordLabels}
          />
        )
      default:
        return <Field {...fieldProps} type="text" />
    }
  }
}

AccountField.propTypes = {
  hasError: PropTypes.bool,
  initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  onKeyUp: PropTypes.func,
  role: PropTypes.string,
  type: PropTypes.oneOf(['date', 'dropdown', 'email', 'password', 'text']),
  t: PropTypes.func
}

// As SelectBox component from Cozy-UI, rendering dropdown type, is just giving
// us the full Option object, we just get its value to facilitate mapping
// with account
const parse = type => value => {
  return type === 'dropdown' ? value.value : value
}

export class AccountFields extends PureComponent {
  render() {
    const {
      container,
      disabled,
      hasError,
      initialValues,
      manifestFields,
      onKeyUp,
      t
    } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div>
        {namedFields.map((field, index) => (
          <FinalFormField
            key={index}
            name={field.name}
            parse={parse(field.type)}
          >
            {({ input }) => (
              <AccountField
                {...field}
                {...input}
                container={container}
                hasError={hasError}
                disabled={disabled}
                initialValue={
                  initialValues[field.name] ||
                  initialValues[getEncryptedFieldName(field.name)]
                }
                onKeyUp={onKeyUp}
                t={t}
              />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

AccountFields.propTypes = {
  disabled: PropTypes.bool,
  hasError: PropTypes.bool,
  fillEncrypted: PropTypes.bool,
  manifestFields: PropTypes.object.isRequired,
  onKeyUp: PropTypes.func,
  t: PropTypes.func
}

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { konnector, lang } = props
    const { locales } = konnector
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  /**
   * Indicates if the state of ReactFinalForm implies that data can be submitted
   * @param  {Object} formState See https://github.com/final-form/final-form#formstate
   * @return {Boolean}
   */
  isSubmittable({ dirty, error, initialValues, valid }) {
    const untouched = initialValues && !dirty
    return error || (valid && !untouched)
  }

  handleKeyUp(event, { dirty, form, initialValues, valid, values }) {
    if (
      event.code === 'Enter' &&
      this.isSubmittable({ dirty, initialValues, valid })
    ) {
      this.handleSubmit(values, form)
    }
  }

  handleSubmit(values, form) {
    const { onSubmit } = this.props
    // Reset form with new values to set back dirty to false
    form.reset(values)
    onSubmit(values)
  }

  validate = (fields, initialValues) => vals => {
    let errors = {}
    for (let name in fields)
      if (
        fields[name].required &&
        !vals[name] &&
        // Don't require value for empty encrypted fields with initial value
        !initialValues[getEncryptedFieldName(name)]
      )
        errors[name] = VALIDATION_ERROR_REQUIRED_FIELD
    return errors
  }

  render() {
    const { account, error, konnector, onSubmit, submitting, t } = this.props
    const { fields, oauth } = konnector

    if (oauth) return <OAuthForm initialValues={initialValues} oauth={oauth} />

    const sanitizedFields = manifest.sanitizeFields(fields)
    const defaultValues = manifest.defaultFieldsValues(sanitizedFields)
    const initialValues = account && account.auth
    const initialAndDefaultValues = { ...defaultValues, ...initialValues }

    let container = null

    return (
      <Form
        initialValues={initialAndDefaultValues}
        onSubmit={onSubmit}
        validate={this.validate(sanitizedFields, initialAndDefaultValues)}
        render={({ dirty, form, values, valid }) => (
          <div
            ref={element => {
              container = element
            }}
          >
            {error && (
              <AccountFormError error={error} konnector={konnector} t={t} />
            )}
            <AccountFields
              container={container}
              disabled={submitting}
              hasError={
                error &&
                error instanceof KonnectorJobError &&
                error.isLoginError()
              }
              initialValues={initialAndDefaultValues}
              manifestFields={sanitizedFields}
              onKeyUp={event =>
                this.handleKeyUp(event, {
                  dirty,
                  form,
                  initialValues,
                  valid,
                  values
                })
              }
              t={t}
            />
            <Button
              busy={submitting}
              className="u-mt-2 u-mb-1-half"
              disabled={
                submitting ||
                !this.isSubmittable({ dirty, error, initialValues, valid })
              }
              extension="full"
              label={t('accountForm.submit.label')}
              onClick={() => this.handleSubmit(values, form)}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  konnector: PropTypes.object.isRequired,
  error: PropTypes.object,
  submitting: PropTypes.bool
}

export default translate()(AccountForm)
