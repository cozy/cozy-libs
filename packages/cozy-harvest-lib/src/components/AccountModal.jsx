import React, { Component } from 'react'
import PropTypes from 'prop-types'

import get from 'lodash/get'
import { withMutations } from 'cozy-client'
import { withRouter } from 'react-router'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import { ModalContent } from 'cozy-ui/transpiled/react/Modal'
import Infos from 'cozy-ui/transpiled/react/Infos'
import Button from 'cozy-ui/transpiled/react/Button'
import { translate } from 'cozy-ui/transpiled/react/I18n'

import accountMutations from '../connections/accounts'
import triggersMutations from '../connections/triggers'
import * as triggersModel from '../helpers/triggers'
import KonnectorConfiguration from './KonnectorConfiguration/KonnectorConfiguration'
import AccountSelectBox from './AccountSelectBox/AccountSelectBox'
import KonnectorModalHeader from './KonnectorModalHeader'

/**
 * AccountModal take an accountId and a list of accounts containing their
 * respecting triggers and display the selected account and the accounts linked
 * to this konnector
 *
 * You have to pass an array of accounts containing their associated trigger:
 * ```
 * accounts: [
 *  {
 *    account: {
 *       _id: '',
 *    },
 *    trigger: {
 *      id: '',
 *      current_state: {
 *      }
 *    }
 *  }
 * ]
 * ```
 *
 * from this array and from the passed accountId, we can fetch the account.
 *
 */
class AccountModal extends Component {
  state = {
    trigger: null,
    account: null,
    fetching: true,
    error: false
  }
  componentDidMount() {
    this.loadSelectedAccountId()
  }

  componentDidUpdate(prevProps) {
    if (this.props.accountId !== prevProps.accountId)
      this.loadSelectedAccountId()
  }

  loadSelectedAccountId() {
    const { accountId, accounts } = this.props
    const matchingTrigger = get(
      accounts.find(account => account.account._id === accountId),
      'trigger'
    )
    if (matchingTrigger) this.fetchAccount(matchingTrigger)
  }

  async fetchAccount(trigger) {
    const { findAccount } = this.props
    this.setState({ fetching: true })

    try {
      const account = await findAccount(triggersModel.getAccountId(trigger))
      this.setState({
        account,
        trigger
      })
      return account
    } catch (error) {
      this.setState({
        error
      })
    } finally {
      this.setState({
        fetching: false
      })
    }
  }

  render() {
    const { konnector, onDismiss, history, accounts, t } = this.props
    const { trigger, account, fetching, error } = this.state
    return (
      <>
        <KonnectorModalHeader konnector={konnector}>
          {account && (
            <AccountSelectBox
              selectedAccount={account}
              accountsList={accounts}
              onChange={option => {
                history.push(`../${option.account._id}`)
              }}
              onCreate={() => {
                history.push(`../../new`)
              }}
            />
          )}
        </KonnectorModalHeader>
        <ModalContent>
          {error && (
            <Infos
              actionButton={
                <Button
                  theme="danger"
                  onClick={this.loadSelectedAccountId.bind(this)}
                  label={t('modal.konnector.error.button')}
                />
              }
              title={t('modal.konnector.error.title')}
              text={t('modal.konnector.error.description', error)}
              icon="warning"
              isImportant
            />
          )}
          {fetching && (
            <div className="u-ta-center">
              <Spinner size="xxlarge" />
            </div>
          )}
          {!error && !fetching && (
            <KonnectorConfiguration
              konnector={konnector}
              trigger={trigger}
              account={account}
              onAccountDeleted={onDismiss}
              addAccount={() => history.push('../new')}
            />
          )}
        </ModalContent>
      </>
    )
  }
}

AccountModal.propTypes = {
  konnector: PropTypes.object.isRequired,
  onDismiss: PropTypes.func,
  history: PropTypes.object.isRequired,
  accounts: PropTypes.array.isRequired,
  t: PropTypes.func.isRequired,
  findAccount: PropTypes.func.isRequired,
  accountId: PropTypes.string.isRequired
}
export default withMutations(accountMutations, triggersMutations)(
  withRouter(translate()(AccountModal))
)
