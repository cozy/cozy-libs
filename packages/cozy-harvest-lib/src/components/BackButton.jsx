import React from 'react'
import Icon from 'cozy-ui/transpiled/react/Icon'
import cx from 'classnames'
import withLocales from './hoc/withLocales'

const BackButton = props => {
  const { className, t, ...rest } = props

  return (
    <button
      className={cx(
        'u-inline-flex u-flex-items-center u-bg-white u-bdw-0 u-ph-0 u-pv-1 u-c-pointer',
        className
      )}
      {...rest}
    >
      <Icon icon="left" className="u-mr-half" />
      {t('Revenir Ã  la liste')}
    </button>
  )
}

export default withLocales(BackButton)
