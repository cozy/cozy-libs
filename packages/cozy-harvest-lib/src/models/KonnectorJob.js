import MicroEE from 'microee'

import accounts from '../helpers/accounts'
import triggers from '../helpers/triggers'

import accountsMutations from '../connections/accounts'
import triggersMutations from '../connections/triggers'

// events
export const ERROR_EVENT = 'error'
export const LOGIN_SUCCESS_EVENT = 'loginSuccess'
export const SUCCESS_EVENT = 'success'
export const TWO_FA_REQUEST_EVENT = 'twoFARequest'
export const TWO_FA_MISMATCH_EVENT = 'twoFAMismatch'

// statuses
export const IDLE = 'IDLE'
export const PENDING = 'PENDING'
export const ERRORED = 'ERRORED'
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS'
export const SUCCESS = 'SUCCESS'
export const RUNNING_TWOFA = 'RUNNING_TWOFA'
export const TWO_FA_MISMATCH = 'TWO_FA_MISMATCH'
export const TWO_FA_REQUEST = 'TWO_FA_REQUEST'

// helpers
export const prepareTriggerAccount = async (trigger, accountsMutations) => {
  const { findAccount, updateAccount } = accountsMutations
  const account = await findAccount(triggers.getAccountId(trigger))
  if (!account) throw new Error('Trigger has no account')
  return await updateAccount(accounts.resetState(account))
}

export class KonnectorJob {
  constructor(client, trigger) {
    this.client = client
    this.trigger = trigger
    this.account = null
    this.unsubscribeAllRealtime = null

    this.accountsMutations = accountsMutations(this.client)
    this.triggersMutations = triggersMutations(this.client)

    // Bind methods used as callbacks
    this.getTwoFACodeProvider = this.getTwoFACodeProvider.bind(this)
    this.getKonnectorSlug = this.getKonnectorSlug.bind(this)
    this.handleLegacyEvent = this.handleLegacyEvent.bind(this)
    this.handleTwoFA = this.handleTwoFA.bind(this)
    this.launch = this.launch.bind(this)
    this.sendTwoFACode = this.sendTwoFACode.bind(this)
    this.unwatch = this.unwatch.bind(this)

    // status and setter/getters
    this._status = IDLE
    this.setStatus = this.setStatus.bind(this)
    this.getStatus = this.getStatus.bind(this)
    this.isRunning = this.isRunning.bind(this)
    this.isTwoFARunning = this.isTwoFARunning.bind(this)
    this.isTwoFARetry = this.isTwoFARetry.bind(this)
  }

  setStatus(status) {
    this._status = status
  }

  getStatus() {
    return this._status
  }

  isRunning() {
    return ![ERRORED, IDLE, SUCCESS].includes(this.getStatus())
  }

  isTwoFARunning() {
    return this.getStatus() === RUNNING_TWOFA
  }

  isTwoFARetry() {
    return this.getStatus() === TWO_FA_MISMATCH
  }

  getTwoFACodeProvider() {
    return accounts.getTwoFACodeProvider(this.account)
  }

  getKonnectorSlug() {
    return this.trigger.message.konnector
  }

  // TODO: Pass updated account as parameter
  handleTwoFA(state) {
    const hasChanged = this.account.state !== state
    if (!hasChanged) return

    this.account.state = state

    if (accounts.isTwoFANeeded(state)) {
      this.setStatus(TWO_FA_REQUEST)
      this.emit(TWO_FA_REQUEST_EVENT, this.account)
    } else if (accounts.isTwoFARetry(state)) {
      this.setStatus(TWO_FA_MISMATCH)
      this.emit(TWO_FA_MISMATCH_EVENT, this.account)
    }
  }

  /**
   * Legacy events use the KonnectorJobWatcher until it has been merged into
   * KonnectorJob so we need to re-emit the events
   */
  handleLegacyEvent(event) {
    return (...args) => {
      switch (event) {
        case ERROR_EVENT:
          this.setStatus(ERRORED)
          break
        case LOGIN_SUCCESS_EVENT:
          this.setStatus(LOGIN_SUCCESS)
          break
        case SUCCESS_EVENT:
          this.setStatus(SUCCESS)
          break
      }
      this.emit(event, ...args)
    }
  }

  /**
   * Send Two FA Code, i.e. save it into account
   */
  async sendTwoFACode(code) {
    this.setStatus(RUNNING_TWOFA)
    const { updateAccount } = this.accountsMutations
    try {
      await updateAccount(accounts.updateTwoFaCode(this.account, code))
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error(error)
      this.setStatus(ERRORED)
    }
  }

  /**
   * Launch the job and set up everything to follow execution.
   */
  async launch() {
    this.setStatus(PENDING)
    const { watchKonnectorAccount } = this.accountsMutations
    const { launchTrigger, watchKonnectorJob } = this.triggersMutations

    this.account = await prepareTriggerAccount(
      this.trigger,
      this.accountsMutations
    )

    const jobWatcher = watchKonnectorJob(await launchTrigger(this.trigger))
    // Temporary reEmitting until merging of KonnectorJobWatcher and
    // KonnectorAccountWatcher into KonnectorJob
    jobWatcher.on(ERROR_EVENT, this.handleLegacyEvent(ERROR_EVENT))
    jobWatcher.on(
      LOGIN_SUCCESS_EVENT,
      this.handleLegacyEvent(LOGIN_SUCCESS_EVENT)
    )
    jobWatcher.on(SUCCESS_EVENT, this.handleLegacyEvent(SUCCESS_EVENT))

    const accountWatcher = watchKonnectorAccount(this.account, {
      onTwoFACodeAsked: this.handleTwoFA,
      onLoginSuccess: jobWatcher.handleSuccess,
      onLoginSuccessHandled: jobWatcher.disableSuccessTimer
    })

    this.unsubscribeAllRealtime = () => {
      jobWatcher.unsubscribeAll()
      accountWatcher.unsubscribeAll()
    }
  }

  unwatch() {
    if (typeof this.unsubscribeAllRealtime === 'function') {
      this.unsubscribeAllRealtime()
    }
  }
}

MicroEE.mixin(KonnectorJob)

export default KonnectorJob
