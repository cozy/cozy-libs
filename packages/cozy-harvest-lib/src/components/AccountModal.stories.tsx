import type { Meta, StoryObj } from '@storybook/react'
import { withRouter } from 'storybook-addon-react-router-v6'

import { AccountModal } from './AccountModal'
import ConfigurationTab from './KonnectorConfiguration/ConfigurationTab'
import { storybookClient } from '../../.storybook/StoryContainer'
import { AccountModalFixtures } from '../../.storybook/fixtures/AccountModal.fixtures'

const meta: Meta<typeof AccountModal> = {
  argTypes: {
    onDismiss: { action: 'onDismiss' },
    replaceHistory: { action: 'replaceHistory' }
  },
  decorators: [withRouter],
  args: {
    client: storybookClient,
    accountId: '1337',
    t: (key: string) => key,
    showAccountSelection: true,
    showNewAccountButton: true,
    accounts: [],
    initialActiveTab: 'configuration',
    Component: ConfigurationTab,
    location: {
      hash: '',
      key: '5j2q3d',
      pathname: '/connected/stub',
      search: '',
      state: null
    },
    accountsAndTriggers: AccountModalFixtures.accountsAndTriggers,
    konnector: AccountModalFixtures.konnector,
    breakpoints: {
      isMobile: false,
      isTablet: false,
      isDesktop: true
    }
  },
  component: AccountModal,
  parameters: {
    reactRouter: {
      routePath: '/connected/stub/accounts/1337/config'
    }
  }
}

export default meta

type Story = StoryObj<typeof AccountModal>

export const WithConfigurationTab: Story = {
  args: {
    Component: ConfigurationTab
  }
}
