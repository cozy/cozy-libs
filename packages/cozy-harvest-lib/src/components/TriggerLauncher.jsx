import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { withClient, withMutations } from 'cozy-client'
import CozyRealtime from 'cozy-realtime'

import TwoFAModal from './TwoFAModal'
import { accountsMutations } from '../connections/accounts'
import { triggersMutations } from '../connections/triggers'
import withLocales from './hoc/withLocales'
import * as triggersModel from '../helpers/triggers'
import * as jobsModel from '../helpers/jobs'
import KonnectorJob, {
  ERROR_EVENT,
  SUCCESS_EVENT,
  LOGIN_SUCCESS_EVENT,
  TWO_FA_REQUEST_EVENT
} from '../models/KonnectorJob'

/**
 * Trigger Launcher renders its children with following props:
 * * launch: Callback to launch the trigger
 * * running: Boolean which indicates if the trigger is running
 * * error: KonnectorJobError
 * * trigger: The current Trigger
 *
 * ### Example
 * ```js
 * <TriggerLauncher initialTrigger={trigger}>
 *   {(launch, running, trigger, error) => (
 *     <Button onClick={() => launch(trigger)} disabled={running} />
 *    )}
 * </TriggerLauncher>
 * ```
 *
 * TODO: inject other props to indicates status (for example: loginSucceed),
 * and use callbacks like `onLoginSuccess` to make this component usable in
 * TriggerManager for example
 *
 */
export class TriggerLauncher extends Component {
  constructor(props, context) {
    super(props, context)
    const { initialTrigger, client } = this.props
    this.state = {
      showTwoFAModal: false,
      trigger: initialTrigger,
      error: triggersModel.getKonnectorJobError(initialTrigger)
    }

    this.dismissTwoFAModal = this.dismissTwoFAModal.bind(this)
    this.displayTwoFAModal = this.displayTwoFAModal.bind(this)
    this.handleError = this.handleError.bind(this)
    this.handleSuccess = this.handleSuccess.bind(this)
    this.handleLoginSuccess = this.handleLoginSuccess.bind(this)

    this.launch = this.launch.bind(this)

    this.realtime = new CozyRealtime({ client })
  }

  async handleUpdate(job) {
    if (this.state.trigger && this.state.trigger._id === job.trigger_id) {
      const trigger = await this.refetchTrigger()
      this.setState({
        trigger,
        running: job.state === 'running', //TODO possible race ?
        error: jobsModel.getKonnectorJobError(job)
      })
    }
  }
  async componentDidMount() {
    await this.realtime.subscribe(
      'updated',
      'io.cozy.jobs',
      this.handleUpdate.bind(this)
    )
  }
  async componentWillUnmount() {
    await this.realtime.unsubscribeAll()
  }
  launch(trigger) {
    const { client, onLaunch } = this.props
    this.konnectorJob = new KonnectorJob(client, trigger)
    this.konnectorJob
      .on(ERROR_EVENT, this.handleError)
      .on(SUCCESS_EVENT, this.handleSuccess)
      .on(LOGIN_SUCCESS_EVENT, this.handleLoginSuccess)
      .on(TWO_FA_REQUEST_EVENT, this.displayTwoFAModal)

    this.setState({
      error: null,
      running: true,
      trigger
    })

    if (typeof onLaunch === 'function') onLaunch(trigger)

    this.konnectorJob.launch()
  }

  dismissTwoFAModal() {
    this.setState({ showTwoFAModal: false })
  }

  displayTwoFAModal() {
    this.setState({ showTwoFAModal: true })
  }

  async handleError(error) {
    /**
     * We don't setState error / running here, since it's the job of `refeshTrigger`
     * to do it
     */
    if (this.state.showTwoFAModal) {
      this.dismissTwoFAModal()
    }
    this.stopWatchingKonnectorJob()

    const { onError } = this.props
    if (typeof onError === 'function') onError(error)
  }

  /**
   * Passing trigger to the callbacks is required for
   * redirecting to the right route after an account creation.
   *
   * We need to do that on both, `handleSuccess` and `handleLoginSuccess`
   * since we can receive a sucess before a loginSuccess (since only
   * few konnectors are dealing with loginSuccess)
   *
   */
  async handleSuccess() {
    if (this.state.showTwoFAModal) {
      this.dismissTwoFAModal()
    }
    this.stopWatchingKonnectorJob()
    const { onSuccess } = this.props
    const { trigger } = this.state
    if (typeof onSuccess === 'function') onSuccess(trigger)
  }

  async handleLoginSuccess() {
    if (this.state.showTwoFAModal) {
      this.dismissTwoFAModal()
    }
    const { onLoginSuccess } = this.props
    const { trigger } = this.state
    if (typeof onLoginSuccess === 'function') onLoginSuccess(trigger)
  }

  async refetchTrigger() {
    const { fetchTrigger } = this.props
    const { trigger } = this.state
    try {
      return await fetchTrigger(trigger._id)
    } catch (error) {
      this.setState({ error, running: false })
      throw error
    }
  }

  stopWatchingKonnectorJob() {
    this.konnectorJob.unwatch()
    delete this.konnectorJob
  }

  render() {
    const { error, running, showTwoFAModal, trigger } = this.state

    const { children, submitting } = this.props
    return (
      <div>
        {children({
          error,
          launch: this.launch,
          running: !!running || !!submitting,
          trigger
        })}
        {showTwoFAModal && (
          <TwoFAModal
            dismissAction={this.dismissTwoFAModal}
            into="coz-harvest-modal-place"
            konnectorJob={this.konnectorJob}
          />
        )}
      </div>
    )
  }
}

TriggerLauncher.propTypes = {
  /**
   * Mandatory for render prop
   */
  children: PropTypes.func.isRequired,
  /**
   * CozyClient instance
   */
  client: PropTypes.object.isRequired,
  /**
   * Callback to call when the trigger is launched
   */
  onLaunch: PropTypes.func,
  /**
   * Callback to call when the konnector job succeeds
   */
  onSuccess: PropTypes.func,
  /**
   * Callback to call when the konnector is logged in
   */
  onLoginSuccess: PropTypes.func,
  /**
   * Callback to call when the konnector job fails
   */
  onError: PropTypes.func,
  /**
   * Indicates if trigger is already runnning.
   * Useful at the initial render of the page when we know a job is already running.
   */
  submitting: PropTypes.bool,
  /**
   * Optionnal trigger in its initial state
   */
  initialTrigger: PropTypes.object
}

export default withLocales(
  withClient(
    withMutations(accountsMutations, triggersMutations)(TriggerLauncher)
  )
)
