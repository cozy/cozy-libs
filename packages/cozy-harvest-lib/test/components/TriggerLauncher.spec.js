import React, { Component } from 'react'
import { shallow } from 'enzyme'

import { TriggerLauncher } from 'components/TriggerLauncher'
import KonnectorJob from 'models/KonnectorJob'

const client = {}
const trigger = {
  _id: '65c347d1ef144288a64105702cc36e59'
}

const props = {
  client,
  trigger
}

class Child extends Component {}

describe('TriggerLauncher', () => {
  // Chainable mock
  const onMock = jest.fn(() => ({
    on: onMock
  }))

  beforeAll(() => {
    jest.spyOn(KonnectorJob.prototype, 'on').mockImplementation(onMock)
    jest.spyOn(KonnectorJob.prototype, 'launch').mockImplementation(() => {})
    jest.spyOn(KonnectorJob.prototype, 'unwatch').mockImplementation(() => {})
  })

  afterEach(() => {
    jest.clearAllMocks()
    onMock.mockClear()
  })

  afterAll(() => {
    jest.resetAllMocks()
  })

  it('should render', () => {
    const component = shallow(
      <TriggerLauncher {...props}>{() => <Child />}</TriggerLauncher>
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should inject launch callback', () => {
    const wrapper = shallow(
      <TriggerLauncher {...props}>
        {({ launch }) => <Child launch={launch} />}
      </TriggerLauncher>
    )

    const childWrapper = wrapper.find(Child)
    const childProps = childWrapper.props()
    expect(typeof childProps.launch).toBe('function')
  })

  it('should inject running prop', () => {
    const wrapper = shallow(
      <TriggerLauncher {...props}>
        {({ running }) => <Child running={running} />}
      </TriggerLauncher>
    )

    const childWrapper = wrapper.find(Child)
    const childProps = childWrapper.props()
    expect(typeof childProps.running).toBe('boolean')
  })

  describe('launch', () => {
    it('should launch KonnectorJob', () => {
      const wrapper = shallow(
        <TriggerLauncher {...props}>
          {({ launch }) => <Child launch={launch} />}
        </TriggerLauncher>
      )

      const childWrapper = wrapper.find(Child)
      const childProps = childWrapper.props()
      childProps.launch()
      expect(KonnectorJob.prototype.launch).toHaveBeenCalled()
    })

    it('should re-render with running being `true`', () => {
      const wrapper = shallow(
        <TriggerLauncher {...props}>
          {({ launch, running }) => <Child launch={launch} running={running} />}
        </TriggerLauncher>
      )

      wrapper.instance().launch()
      wrapper.update()
      const childWrapper = wrapper.find(Child)
      expect(childWrapper.props().running).toBe(true)
    })

    it('should register to KonnectorJob events', () => {
      const wrapper = shallow(
        <TriggerLauncher {...props}>
          {({ launch, running }) => <Child launch={launch} running={running} />}
        </TriggerLauncher>
      )

      let childWrapper = wrapper.find(Child)
      childWrapper.props().launch()
      expect(onMock).toHaveBeenCalledWith(
        'error',
        wrapper.instance().handleError
      )
      expect(onMock).toHaveBeenCalledWith(
        'success',
        wrapper.instance().handleSuccess
      )
      expect(onMock).toHaveBeenCalledWith(
        'twoFARequest',
        wrapper.instance().handleTwoFA
      )
      expect(onMock).toHaveBeenCalledWith(
        'twoFARequest',
        wrapper.instance().handleTwoFA
      )
    })

    it('should clear Error', () => {
      const wrapper = shallow(
        <TriggerLauncher {...props}>
          {({ error, launch, running }) => (
            <Child error={error} launch={launch} running={running} />
          )}
        </TriggerLauncher>
      )

      wrapper.instance().launch()
      wrapper.update()
      wrapper.instance().handleError(new Error('Test error'))
      wrapper.update()
      // Relaunch
      wrapper.instance().launch()
      wrapper.update()
      const childWrapper = wrapper.find(Child)
      expect(childWrapper.props().error).toBe(null)
    })
  })

  describe('handleError', () => {
    it('should re-render with running being `false`', () => {
      const wrapper = shallow(
        <TriggerLauncher {...props}>
          {({ error, launch, running }) => (
            <Child error={error} launch={launch} running={running} />
          )}
        </TriggerLauncher>
      )

      wrapper.instance().launch()
      wrapper.update()
      wrapper.instance().handleError(new Error('Test error'))
      wrapper.update()
      const childWrapper = wrapper.find(Child)
      expect(childWrapper.props().running).toBe(false)
    })

    it('should re-render with error being defined', () => {
      const wrapper = shallow(
        <TriggerLauncher {...props}>
          {({ error, launch, running }) => (
            <Child error={error} launch={launch} running={running} />
          )}
        </TriggerLauncher>
      )

      const error = new Error('Test error')

      wrapper.instance().launch()
      wrapper.update()
      wrapper.instance().handleError(error)
      wrapper.update()
      const childWrapper = wrapper.find(Child)
      expect(childWrapper.props().error).toEqual(error)
    })

    it('should hide twoFA modal', () => {
      const wrapper = shallow(
        <TriggerLauncher {...props}>
          {({ launch, running }) => <Child launch={launch} running={running} />}
        </TriggerLauncher>
      )

      wrapper.instance().launch()
      wrapper.update()
      wrapper.instance().displayTwoFAModal()
      wrapper.instance().handleError(new Error('Test error'))
      wrapper.update()
      expect(wrapper.getElement()).toMatchSnapshot()
    })

    it('should unsubscribe from KonnectorJob', () => {
      const wrapper = shallow(
        <TriggerLauncher {...props}>
          {({ launch, running }) => <Child launch={launch} running={running} />}
        </TriggerLauncher>
      )

      wrapper.instance().launch()
      wrapper.update()
      wrapper.instance().displayTwoFAModal()
      wrapper.instance().handleError(new Error('Test error'))
      wrapper.update()
      expect(wrapper.getElement()).toMatchSnapshot()
    })
  })

  describe('handleSuccess', () => {
    it('should re-render with running being `false`', () => {
      const wrapper = shallow(
        <TriggerLauncher {...props}>
          {({ launch, running }) => <Child launch={launch} running={running} />}
        </TriggerLauncher>
      )

      wrapper.instance().launch()
      wrapper.update()
      wrapper.instance().handleSuccess()
      wrapper.update()
      const childWrapper = wrapper.find(Child)
      expect(childWrapper.props().running).toBe(false)
    })

    it('should hide twoFA modal', () => {
      const wrapper = shallow(
        <TriggerLauncher {...props}>
          {({ launch, running }) => <Child launch={launch} running={running} />}
        </TriggerLauncher>
      )

      wrapper.instance().launch()
      wrapper.instance().displayTwoFAModal()
      wrapper.instance().handleSuccess()
      wrapper.update()
      expect(wrapper.getElement()).toMatchSnapshot()
    })
  })

  describe('handleTwoFA', () => {
    it('should display Two FA modal', () => {
      const wrapper = shallow(
        <TriggerLauncher {...props}>
          {({ launch, running }) => <Child launch={launch} running={running} />}
        </TriggerLauncher>
      )

      wrapper.instance().handleTwoFA()
      expect(wrapper.getElement()).toMatchSnapshot()
    })
  })
})
