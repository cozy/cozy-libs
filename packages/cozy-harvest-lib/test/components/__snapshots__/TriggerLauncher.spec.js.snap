// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TriggerLauncher handleError should hide twoFA modal 1`] = `
<div>
  <Child
    launch={[Function]}
    running={false}
  />
</div>
`;

exports[`TriggerLauncher handleError should unsubscribe from KonnectorJob 1`] = `
<div>
  <Child
    launch={[Function]}
    running={false}
  />
</div>
`;

exports[`TriggerLauncher handleSuccess should hide twoFA modal 1`] = `
<div>
  <Child
    launch={[Function]}
    running={false}
  />
</div>
`;

exports[`TriggerLauncher handleTwoFA should display Two FA modal 1`] = `
<div>
  <Child
    launch={[Function]}
    running={false}
  />
  <Wrapper
    dismissAction={[Function]}
    into="coz-harvest-modal-place"
  />
</div>
`;

exports[`TriggerLauncher should render 1`] = `
<div>
  <Child />
</div>
`;
